File: Contingency table.py
# Table of Contigency
# Contingency table also known as cross tabulation or crosstab is a tabular mechanism 
# in matrix format that display multivariate frequency distribution of the variables. 
# This table is understood as the relationship between categorical variables.

import numpy as np
import pandas as pd
import matplotlib as plt

# Define a dictionary containing data
data = {'fruit':['Orange', 'Banana', 'Orange', 'Banana'],
        'like':['yes', 'yes', 'no', 'no'],
        'count':[22, 16, 406, 841]
}
 
# Convert the dictionary into DataFrame 
df = pd.DataFrame(data)
print(df)
#     fruit like  count
# 0  Orange  yes     22
# 1  Banana  yes     16
# 2  Orange   no    406
# 3  Banana   no    841

# Displaying information of dataframe
df.info()
# <class 'pandas.core.frame.DataFrame'>
# RangeIndex: 4 entries, 0 to 3
# Data columns (total 3 columns):
#   Column  Non-Null Count  Dtype
# ---  ------  --------------  -----
# 0   fruit   4 non-null      object
# 1   like    4 non-null      object
# 2   count   4 non-null      int64
# dtypes: int64(1), object(2)
# memory usage: 224.0+ bytes

# Displaying data type information
df.dtypes
# fruit    object
# like     object
# count     int64
# dtype: object

#Getting and Displaying Cross Table
df_ctab = pd.crosstab(df['fruit'], data['like'], margins = True)
print(df_ctab)

File: Contingency table.r
# Table of Contigency
# Contingency table also known as cross tabulation or crosstab is a tabular mechanism 
# in matrix format that display multivariate frequency distribution of the variables. 
# This table is understood as the relationship between categorical variables.

fruit <- c("Orange", "Banana")
like <- c("yes","no")
ocurrancy <- c(22,16,406,841)

table <- expand.grid(fruit=fruit,
                     like=like
                     )
# table
# fruit like
# 1 Orange  yes
# 2 Banana  yes
# 3 Orange   no
# 4 Banana   no

table <- cbind(table, count=ocurrancy)
table2 <- xtabs(ocurrancy ~ fruit+like,
      data=table
      )
# like
# fruit    yes  no
# Orange  22 406
# Banana  16 841

n1 <- sum(table2[1,])
n2 <- sum(table2[2,])
p1 <- table2[1,1]/n1
p2 <- table2[2,1]/n2

alpha <- 0.05
zalfa_dif <-qnorm(1-alpha/2)

# Differences
dif <- p1-p2
sum1 <- (p1*(1-p1))/n1  
sum2 <- (p2*(1-p2))/n2
dev_dif <- sqrt(sum1+sum2)

int_dif <- c(dif-zalfa_dif*dev_dif, 
             dif+zalfa_dif*dev_dif)

#Odds Ratio
risk <- p1/p2
sum1r <- (1-p1)/(n1*p1)
sum2r <- (1-p2)/(n2*p2)
dev_lr <- sqrt(sum1r+sum2r)
int_lr <- log(risk) + c(-1,1)*zalfa_dif*dev_lr
int_r <- exp(int_lr)

or <- (p1/(1-p1))/(p2/(1-p2))
dev_lor <- sqrt(sum(1/table2))
int_lor <- log(or) + c(-1,1)*zalfa_dif*dev_lor
int_or <- exp(int_lor)

